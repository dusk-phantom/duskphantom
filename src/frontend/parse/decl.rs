use winnow::error::{ErrMode, ErrorKind};

use super::*;

pub fn make_const(decl: Decl) -> Decl {
    match decl {
        Decl::Var(ty, id, expr) => Decl::Const(ty, id, expr),
        _ => decl,
    }
}

pub fn decl(input: &mut &str) -> PResult<Decl> {
    // Attempt to match a macro.
    if input.starts_with('#') {
        return alt((
            (token("#include"), opt(take_until(0.., '\n')), blank).value(Decl::Stack(vec![])),
            (token("#define"), pad(ident), expr)
                .map(|(_, id, expr)| Decl::Const(Type::Int, id, Some(expr))),
        ))
        .parse_next(input);
    }

    // Match const token.
    let is_const = opt(token("const")).parse_next(input)?.is_some();

    // Consume extern token.
    opt(token("extern")).parse_next(input)?;

    // Parse type.
    let left_type = atom_type.parse_next(input)?;

    // Parse lval and optional assignment expression.
    let mut decls: Vec<Decl> = separated(
        1..,
        |input: &mut &str| assignment(input, left_type.clone()),
        token(","),
    )
    .parse_next(input)?;

    // Require semicolon if the last declaration is not function implementation
    if let Some(Decl::Func(_, _, Some(_))) = decls.last() {
        // Do nothing
    } else {
        token(";").parse_next(input)?;
    }

    // Make constant if necessary
    if is_const {
        decls = decls.into_iter().map(make_const).collect();
    }

    // Return declaration according to count
    match decls.len() {
        1 => Ok(decls.pop().unwrap()),
        _ => Ok(Decl::Stack(decls)),
    }
}

pub fn assignment(input: &mut &str, left_type: Type) -> PResult<Decl> {
    let left_val = lval.parse_next(input)?;
    let typed_ident = acc_lval(left_type, left_val);
    let Some(id) = typed_ident.id else {
        return Err(ErrMode::from_error_kind(input, ErrorKind::Verify).cut());
    };

    // Parse optional assignment.
    if let Some(expr) = opt(preceded(token("="), expr)).parse_next(input)? {
        return Ok(Decl::Var(typed_ident.ty, id, Some(expr)));
    };

    // Parse optional function implementation.
    if let Some(body) = opt(curly(vec_stmt)).parse_next(input)? {
        return Ok(Decl::Func(
            typed_ident.ty,
            id,
            Some(Box::new(Stmt::Block(body))),
        ));
    };

    // Return declaration according to real type
    match typed_ident.ty {
        Type::Function(_, _) => Ok(Decl::Func(typed_ident.ty, id, None)),
        _ => Ok(Decl::Var(typed_ident.ty, id, None)),
    }
}
