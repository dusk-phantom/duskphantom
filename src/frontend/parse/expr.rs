use crate::gen_lrec_binary;

use super::*;

/// Parse a vector of Expr.
pub fn vec_expr(input: &mut &str) -> PResult<Vec<Expr>> {
    separated(0.., expr, token(",")).parse_next(input)
}

/// Parse a box of Expr.
/// Like an `expr`, but returns the boxed version.
pub fn box_expr(input: &mut &str) -> PResult<Box<Expr>> {
    expr.map(Box::new).parse_next(input)
}

/// Parse prefix expressions like `!x->y.z`.
pub fn prefix(input: &mut &str) -> PResult<Expr> {
    let disp = dispatch! { peek(any);
        '{' => curly(separated(0.., expr, token(","))).map(Expr::Array),
        '.' | '0'..='9' => pad(constant_number),
        '"' => pad(string_lit).map(Expr::String),
        'f' => token("false").value(Expr::Int(0)),
        't' => token("true").value(Expr::Int(1)),
        '(' => paren(expr),
        _ => fail,
    };
    let atom = alt((disp, pad(ident).map(Expr::Var)));

    // Postfix: `head[a].b(c)->d`.
    // Tail parsers return mutation on `head`.
    // Closures should be wrapped in `BoxF` for equal sizes.
    // Wrapping all closures with `BoxF` can also fix type inference problems,
    // because all closures have unique types, making `alt` report errors.
    let postfix_tail = dispatch! { peek(any);
        '[' => bracket(box_expr).map(|x| BoxF::new(|acc| Expr::Index(acc, x))),
        '(' => paren(vec_expr).map(|x| BoxF::new(|acc| Expr::Call(acc, x))),
        _ => fail,
    };
    let postfix = lrec(atom, repeat(0.., postfix_tail));

    // Prefix unary operator.
    let prefix_init = unary_op.map(|op| BoxF::new(|acc| Expr::Unary(op, acc)));
    rrec(repeat(0.., prefix_init), postfix).parse_next(input)
}

// Generate parser for each level of expressions,
// featuring binary operators.
gen_lrec_binary!(binary_lv0, binary_op_lv0, prefix);
gen_lrec_binary!(binary_lv1, binary_op_lv1, binary_lv0);
gen_lrec_binary!(binary_lv2, binary_op_lv2, binary_lv1);
gen_lrec_binary!(binary_lv3, binary_op_lv3, binary_lv2);
gen_lrec_binary!(binary_lv4, binary_op_lv4, binary_lv3);
gen_lrec_binary!(binary_lv5, binary_op_lv5, binary_lv4);
gen_lrec_binary!(binary_lv6, binary_op_lv6, binary_lv5);
gen_lrec_binary!(binary_lv7, binary_op_lv7, binary_lv6);
gen_lrec_binary!(binary_lv8, binary_op_lv8, binary_lv7);
gen_lrec_binary!(binary_lv9, binary_op_lv9, binary_lv8);

/// Parse a conditional expression.
pub fn expr(input: &mut &str) -> PResult<Expr> {
    binary_lv9.parse_next(input)
}
