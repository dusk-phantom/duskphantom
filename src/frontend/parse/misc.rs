use super::*;

pub fn map_entry(input: &mut &str) -> PResult<MapEntry> {
    (ident, token(":"), expr)
        .map(|(id, _, expr)| MapEntry::new(id, expr))
        .parse_next(input)
}
