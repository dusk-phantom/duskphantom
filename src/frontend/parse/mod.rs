mod decl;
mod expr;
#[macro_use]
mod macros;
mod misc;
mod oprt;
mod parser;
pub mod program;
mod stmt;
mod typed;

// Re-export parsers for convenient use
pub use self::decl::*;
pub use self::expr::*;
pub use self::misc::*;
pub use self::oprt::*;
pub use self::parser::*;
pub use self::stmt::*;
pub use self::typed::*;

// Re-export structs in "ir"
pub use super::ir::decl::*;
pub use super::ir::expr::*;
pub use super::ir::misc::*;
pub use super::ir::oprt::*;
pub use super::ir::program::*;
pub use super::ir::stmt::*;
pub use super::ir::typed::*;

// Re-export winnow parsers
pub use winnow::ascii::digit1;
pub use winnow::ascii::multispace0;
pub use winnow::ascii::space0;
pub use winnow::ascii::space1;
pub use winnow::combinator::alt;
pub use winnow::combinator::cut_err;
pub use winnow::combinator::empty;
pub use winnow::combinator::fail;
pub use winnow::combinator::opt;
pub use winnow::combinator::peek;
pub use winnow::combinator::preceded;
pub use winnow::combinator::repeat;
pub use winnow::combinator::separated;
pub use winnow::combinator::terminated;
pub use winnow::combinator::trace;
pub use winnow::dispatch;
pub use winnow::error::ContextError;
pub use winnow::error::ParserError;
pub use winnow::error::StrContext;
pub use winnow::stream::AsChar;
pub use winnow::stream::Compare;
pub use winnow::stream::SliceLen;
pub use winnow::stream::Stream;
pub use winnow::stream::StreamIsPartial;
pub use winnow::token::any;
pub use winnow::token::literal;
pub use winnow::token::one_of;
pub use winnow::token::take_until;
pub use winnow::token::take_while;
pub use winnow::PResult;
pub use winnow::Parser;
