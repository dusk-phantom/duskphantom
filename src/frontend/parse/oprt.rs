use super::*;

pub fn unary_op(input: &mut &str) -> PResult<UnaryOp> {
    dispatch! { peek(any);
        '!' => token("!").value(UnaryOp::Not),
        '-' => token("-").value(UnaryOp::Neg),
        '+' => token("+").value(UnaryOp::Pos),
        _ => fail,
    }
    .parse_next(input)
}

/// Level 0 operators, left to right
pub fn binary_op_lv0(input: &mut &str) -> PResult<BinaryOp> {
    dispatch! { peek(any);
        '*' => token("*").value(BinaryOp::Mul),
        '/' => token("/").value(BinaryOp::Div),
        '%' => token("%").value(BinaryOp::Mod),
        _ => fail,
    }
    .parse_next(input)
}

/// Level 1 operators, left to right
pub fn binary_op_lv1(input: &mut &str) -> PResult<BinaryOp> {
    alt((
        token("+").value(BinaryOp::Add),
        token("-").value(BinaryOp::Sub),
    ))
    .parse_next(input)
}

/// Level 2 operators, left to right
pub fn binary_op_lv2(input: &mut &str) -> PResult<BinaryOp> {
    alt((
        token(">>").value(BinaryOp::Shr),
        token("<<").value(BinaryOp::Shl),
    ))
    .parse_next(input)
}

/// Level 3 operators, left to right
pub fn binary_op_lv3(input: &mut &str) -> PResult<BinaryOp> {
    dispatch! { peek(any);
        '>' => alt((
            token(">=").value(BinaryOp::Ge),
            token(">").value(BinaryOp::Gt),
        )),
        '<' => alt((
            token("<=").value(BinaryOp::Le),
            token("<").value(BinaryOp::Lt),
        )),
        _ => fail,
    }
    .parse_next(input)
}

/// Level 4 operators, left to right
pub fn binary_op_lv4(input: &mut &str) -> PResult<BinaryOp> {
    alt((
        token("==").value(BinaryOp::Eq),
        token("!=").value(BinaryOp::Ne),
    ))
    .parse_next(input)
}

/// Level 5 operators, left to right
pub fn binary_op_lv5(input: &mut &str) -> PResult<BinaryOp> {
    token("&").value(BinaryOp::BitAnd).parse_next(input)
}

/// Level 6 operators, left to right
pub fn binary_op_lv6(input: &mut &str) -> PResult<BinaryOp> {
    token("^").value(BinaryOp::BitXor).parse_next(input)
}

/// Level 7 operators, left to right
pub fn binary_op_lv7(input: &mut &str) -> PResult<BinaryOp> {
    token("|").value(BinaryOp::BitOr).parse_next(input)
}

/// Level 8 operators, left to right
pub fn binary_op_lv8(input: &mut &str) -> PResult<BinaryOp> {
    token("&&").value(BinaryOp::And).parse_next(input)
}

/// Level 9 operators, left to right
pub fn binary_op_lv9(input: &mut &str) -> PResult<BinaryOp> {
    token("||").value(BinaryOp::Or).parse_next(input)
}
