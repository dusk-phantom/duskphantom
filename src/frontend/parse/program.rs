use crate::errors::FrontendError;

use super::*;

pub fn parse(src: &str) -> Result<Program, FrontendError> {
    preceded(blank, repeat(0.., decl))
        .map(Program::new)
        .parse(src)
        .map_err(|err| FrontendError::ParseError(err.to_string()))
}
