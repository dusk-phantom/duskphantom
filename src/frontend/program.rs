use super::*;

/// The full program.
/// A executable program is a set of modules with an entry module.
/// For now, only one module is supported, so the only module is entry.
#[derive(Clone, PartialEq, Debug)]
pub struct Program {
    /// The module of the program.
    /// Currently only one module is supported.
    pub module: Module,
}

impl Program {
    pub fn new(decls: Vec<Decl>) -> Self {
        Self { module: decls }
    }
}

/// A module is a single file.
/// Only declaration can appear at top level.
pub type Module = Vec<Decl>;

pub fn parse(src: &str) -> Result<Program, FrontendError> {
    preceded(blank, repeat(0.., decl))
        .map(Program::new)
        .parse(src)
        .map_err(|err| FrontendError::ParseError(err.to_string()))
}

#[allow(unused)]
pub fn optimize(program: &mut Program) {}
