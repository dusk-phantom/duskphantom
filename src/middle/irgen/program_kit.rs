// Copyright 2024 Duskphantom Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

use anyhow::Result;

use crate::middle::ir::FunPtr;
use crate::middle::irgen::value::Value;
use crate::utils::frame_map::FrameMap;
use crate::{frontend, middle};

/// Kit for translating a program to middle IR
pub struct ProgramKit<'a> {
    pub env: FrameMap<'a, String, Value>,
    pub fun_env: FrameMap<'a, String, FunPtr>,
    pub program: &'a mut middle::Program,
}

/// A program kit (top level) can generate declarations
impl<'a> ProgramKit<'a> {
    pub fn gen(mut self, program: &frontend::Program) -> Result<()> {
        self.gen_library_function();
        for decl in program.module.iter() {
            self.gen_global_decl(decl)?;
        }
        for decl in program.module.iter() {
            self.gen_impl(decl)?;
        }
        Ok(())
    }
}
