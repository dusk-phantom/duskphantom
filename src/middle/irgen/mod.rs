use crate::{frontend, middle};
use anyhow::Result;
use program_kit::ProgramKit;
use std::collections::HashMap;

mod constant;
mod function_kit;
mod gen_binary;
mod gen_const_binary;
mod gen_const_expr;
mod gen_const_unary;
mod gen_expr;
mod gen_global_decl;
mod gen_impl;
mod gen_inner_decl;
mod gen_library_function;
mod gen_stmt;
mod gen_unary;
mod program_kit;
mod value;
mod value_type;

/// Generate middle IR from a frontend AST
pub fn gen(program: &frontend::Program) -> Result<middle::Program> {
    let mut result = middle::Program::new();
    ProgramKit {
        program: &mut result,
        env: HashMap::new(),
        fun_env: HashMap::new(),
    }
    .gen(program)?;
    Ok(result)
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use super::*;
    use crate::frontend::program::parse;

    #[test]
    fn test_normal() {
        let code = r#"
            int main() {
                int a = 1;
                int b = 2;
                int c = a + b;
                return c;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_if() {
        let code = r#"
            int main() {
                int a = 1;
                int b = 2;
                if (a < b) {
                    a = 3;
                } else {
                    a = 4;
                }
                return a;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_while() {
        let code = r#"
            int main() {
                int a = 0;
                while (a < 10) {
                    a = a + 1;
                }
                return a;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_do_while() {
        let code = r#"
            int main() {
                int a = 0;
                do {
                    a = a + 1;
                } while (a < 10);
                return a;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_break() {
        let code = r#"
            int main() {
                int a = 0;
                while (a < 10) {
                    a = a + 1;
                    break;
                }
                return a;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_continue() {
        let code = r#"
            int main() {
                int a = 0;
                while (a < 10) {
                    a = a + 1;
                    continue;
                }
                return a;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_control_flow() {
        let code = r#"
            int main() {
                int a = 0;
                while (a < 10) {
                    a = a + 1;
                    if (a == 5) {
                        break;
                    } else {
                        continue;
                    }
                }
                return a;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_default_exit() {
        let code = r#"
            int main() {
                int a = 0;
                while (a < 10) {
                    a = a + 1;
                    if (a == 5) {
                        break;
                    }
                }
                return a;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_dead_code() {
        let code = r#"
            int main() {
                while (1) {
                    break;
                    continue;
                    return 0;
                    break;
                    return 0;
                    continue;
                    continue;
                    return 1;
                    break;
                }
                return 0;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_global_variable() {
        let code = r#"
            int x = 4;
            int y = 8;
            int main() {
                x = x + y;
                return x;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_conv() {
        let code = r#"
            int main() {
                int x = 1;
                float y = 2.0;
                float z = x + y;
                return z;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_zext() {
        let code = r#"
            int main() {
                return (3 > 1) + (4 > 2);
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_param() {
        let code = r#"
            int main(int arg) {
                return arg;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_call() {
        let code = r#"
            int main() {
                return f(1.7);
            }

            int f(int x) {
                return x + 1;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_nested_call() {
        let code = r#"
            int main() {
                return f(f(1));
            }

            int f(int x) {
                return x + 1;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_constant() {
        let code = r#"
            const float PI = 3.1415926;

            int main() {
                return PI;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_constant_array() {
        let code = r#"
            const float A[3][2][2] = {{1}, 1, 4, 5, 1, {4}};

            int main() {
                return A[0][0][0];
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_variable_array() {
        let code = r#"
            int main() {
                float A[2][2][2] = {1, 1, 4, 5, {{1}, 4}};
                return A[1][1][1];
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_assign_array() {
        let code = r#"
            int main() {
                int A[1] = {0};
                A[A[0]] = 1;
                return A[0];
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_pointer() {
        let code = r#"
            int main() {
                int arr[1] = {8};
                f(arr);
                putarray(1, arr);
                return 0;
            }

            int f(int a[]) {
                a[0] = 1;
                return a[0];
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_number_condition() {
        let code = r#"
            int main() {
                float a = 5.4;
                int b = 8;
                int z = 0;
                if (a) {
                    z = 1;
                }
                if (b) {
                    z = 2;
                }
                return z;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_unary() {
        let code = r#"
            int main() {
                int x = 1;
                return !+-x;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_cmp() {
        let code = r#"
            int main() {
                bool x = 1 < 2;
                bool y = 1 < 1.1;
                return x && y;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_constant_fold() {
        let code = r#"
            const int x = 1 + 3;
            const int y = x * x;
            int main() {
                return x + y;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_library_function() {
        let code = r#"
            int main() {
                int x = getint();
                putint(x + 3);
                return 0;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_lazy_eval() {
        let code = r#"
            int main() {
                int x = getint();
                (x > 1) && f(x);
                return 0;
            }

            int f(int x) {
                putint(x);
                return x;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_lazy_eval_with_if() {
        let code = r#"
            int main() {
                int x = getint();
                if (x > 1 && x < 3) {
                    putint(x);
                }
                return 0;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }

    #[test]
    fn test_putf() {
        let code = r#"
            int main() {
                int x = getint();
                putf("x = %d", x);
                return 0;
            }
        "#;
        let program = parse(code).unwrap();
        let result = gen(&program).unwrap();
        let llvm_ir = result.module.gen_llvm_ir();
        assert_snapshot!(llvm_ir);
    }
}
