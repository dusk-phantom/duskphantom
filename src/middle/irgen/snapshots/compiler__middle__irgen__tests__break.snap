---
source: src/middle/irgen/mod.rs
expression: llvm_ir
---
define i32 @main() {
%entry:
%alloca_1 = alloca i32
store i32 0, ptr %alloca_1
br label %cond0

%cond0:
%load_11 = load i32, ptr %alloca_1
%icmp_12 = icmp slt i32 %load_11, 10
br i1 %icmp_12, label %body1, label %final2

%body1:
%load_7 = load i32, ptr %alloca_1
%Add_8 = add i32, %load_7, 1
store i32 %Add_8, ptr %alloca_1
br label %final2

%final2:
%load_14 = load i32, ptr %alloca_1
ret i32 %load_14


}
